test/storage/sql_table_test.cpp

```diff
@@ -246,105 +316,677 @@ class SqlTableRW {
+  storage::layout_version_t version_;
+  storage::SqlTable *table_ = nullptr;
 
+ private:
   storage::BlockStore block_store_{100, 100};
   catalog::table_oid_t table_oid_;
-  storage::SqlTable *table_ = nullptr;
 
+  // keep meta-data for most recent schema
   catalog::Schema *schema_ = nullptr;
   std::vector<catalog::Schema::Column> cols_;
   std::vector<catalog::col_oid_t> col_oids_;
 
   storage::ProjectedRowInitializer *pri_ = nullptr;
   storage::ProjectionMap *pr_map_ = nullptr;
 
-  byte *insert_buffer_ = nullptr;
-  storage::ProjectedRow *insert_ = nullptr;
+  byte *buffer_ = nullptr;
+  storage::ProjectedRow *pr_ = nullptr;
+
+  storage::layout_version_t next_version_ = storage::layout_version_t(0);
+
+  void ResetProjectedRow(const std::vector<catalog::col_oid_t> &col_oids) {
+    auto pr_pair = table_->InitializerForProjectedRow(col_oids, version_);
+    buffer_ = common::AllocationUtil::AllocateAligned(pr_pair.first.ProjectedRowSize());
+    pr_ = pr_pair.first.InitializeRow(buffer_);
+
+    delete pri_;
+    delete pr_map_;
+    pri_ = new storage::ProjectedRowInitializer(std::get<0>(pr_pair));
+    pr_map_ = new storage::ProjectionMap(std::get<1>(pr_pair));
+  }
 };
 
 struct SqlTableTests : public TerrierTest {
   void SetUp() override { TerrierTest::SetUp(); }
 
   void TearDown() override { TerrierTest::TearDown(); }
+
+  storage::RecordBufferSegmentPool buffer_pool_{100, 100};
+  transaction::TransactionManager txn_manager_ = {&buffer_pool_, true, LOGGING_DISABLED};
 };
 
 // NOLINTNEXTLINE
-TEST_F(SqlTableTests, SelectInsertTest) {
-  SqlTableRW table(catalog::table_oid_t(2));
-
+TEST_F(SqlTableTests, SelectTest) {
+  SqlTableTestRW table(catalog::table_oid_t(2));
+  auto txn = txn_manager_.BeginTransaction();
   table.DefineColumn("id", type::TypeId::INTEGER, false, catalog::col_oid_t(0));
   table.DefineColumn("datname", type::TypeId::INTEGER, false, catalog::col_oid_t(1));
   table.Create();
-  table.StartRow();
-  table.SetIntColInRow(0, 100);
-  table.SetIntColInRow(1, 15721);
-  storage::TupleSlot row1_slot = table.EndRowAndInsert();
+  table.StartInsertRow();
+  table.SetIntColInRow(catalog::col_oid_t(0), 100);
+  table.SetIntColInRow(catalog::col_oid_t(1), 10000);
+  storage::TupleSlot row1_slot = table.EndInsertRow(txn);
+
+  table.StartInsertRow();
+  table.SetIntColInRow(catalog::col_oid_t(0), 200);
+  table.SetIntColInRow(catalog::col_oid_t(1), 10001);
+  storage::TupleSlot row2_slot = table.EndInsertRow(txn);
+
+  uint32_t id = table.GetIntColInRow(txn, catalog::col_oid_t(0), row1_slot);
+  EXPECT_EQ(100, id);
+  uint32_t datname = table.GetIntColInRow(txn, catalog::col_oid_t(1), row1_slot);
+  EXPECT_EQ(10000, datname);
 
-  table.StartRow();
-  table.SetIntColInRow(0, 200);
-  table.SetIntColInRow(1, 25721);
-  storage::TupleSlot row2_slot = table.EndRowAndInsert();
+  // manually set the version of the transaction to be 1
+  table.version_ = storage::layout_version_t(1);
```

Comment:
Directly modifying this field is confusing and seems unsafe (may violate your invariant). A method for setting the version of the transaction may better express your intention.

Reply:
The comment is incorrect.  This is not manually setting the transaction's version but rather simulating the catalog update that occurs along with a schema update.  I'll update the comment to reflect this.